src/pip/_internal/models/installation_report.py

```diff
@@ -0,0 +1,57 @@
+from typing import Any, Dict, Sequence
+
+from pip._vendor.packaging.markers import default_environment
+from pip._vendor.packaging.utils import canonicalize_name
+
+from pip import __version__
+from pip._internal.req.req_install import InstallRequirement
+
+
+class InstallationReport:
+    def __init__(self, install_requirements: Sequence[InstallRequirement]):
+        self._install_requirements = install_requirements
+
+    @classmethod
+    def _install_req_to_dict(cls, ireq: InstallRequirement) -> Dict[str, Any]:
+        assert ireq.download_info, f"No download_info for {ireq}"
+        res = {
+            # PEP 610 json for the download URL. download_info.archive_info.hash may
+            # be absent when the requirement was installed from the wheel cache
+            # and the cache entry was populated by an older pip version that did not
+            # record origin.json.
+            "download_info": ireq.download_info.to_dict(),
+            # is_direct is true if the requirement was a direct URL reference (which
+            # includes editable requirements), and false if the requirement was
+            # downloaded from a PEP 503 index or --find-links.
+            "is_direct": bool(ireq.original_link),
+            # requested is true if the requirement was specified by the user (aka
+            # top level requirement), and false if it was installed as a dependency of a
+            # requirement. https://peps.python.org/pep-0376/#requested
+            "requested": ireq.user_supplied,
```

Comment:
These fields are not strictly needed, right? They can be inferred by other entries’ `requires_dist`. I’m not opposed to adding them (they are convinient, if nothing else), just checking I’m not missing anything.

Reply:
They are needed. `is_direct` is important to know if top level requirements were provided as direct URLs, or if dependencies provided as name were constrained to direct URLs. `requested` is important to identify top level (user supplied) requirements that can also be dependencies of other indirect requirements.